src/include/denc.h

```diff
@@ -0,0 +1,1176 @@
+    byte = (v & 0x7f);
+    v >>= 7;
+  }
+  *(__u8*)p.get_ptr_add(1) = byte;
+}
+
+template<typename T>
+inline void denc_varint(T& v, bufferptr::iterator& p) {
+  uint8_t byte = *(__u8*)p.c_str_add(1);
+  v = byte & 0x7f;
+  int shift = 7;
+  while (byte & 0x80) {
+    byte = *(__u8*)p.c_str_add(1);
+    v |= (T)(byte & 0x7f) << shift;
+    shift += 7;
+  }
+}
+
+
+// signed varint encoding
+//
+// low bit = 1 = negative, 0 = positive
+// high bit of every byte indicates whether another byte follows.
+inline void denc_signed_varint(int64_t v, size_t& p) {
+  p += sizeof(v) + 2;
+}
+inline void denc_signed_varint(int64_t v, bufferlist::contiguous_appender& p) {
+  if (v < 0) {
+    v = (-v << 1) | 1;
+  } else {
+    v <<= 1;
+  }
+  denc_varint(v, p);
+}
+
+template<typename T>
+inline void denc_signed_varint(T& v, bufferptr::iterator& p)
+{
+  int64_t i;
+  denc_varint(i, p);
+  if (i & 1) {
+    v = -(i >> 1);
+  } else {
+    v = i >> 1;
+  }
+}
+
+// varint + lowz encoding
+//
+// first(low) 2 bits = how many low zero bits (nibbles)
+// high bit of each byte = another byte follows
+// (so, 5 bits data in first byte, 7 bits data thereafter)
+inline void denc_varint_lowz(uint64_t v, size_t& p) {
+  p += sizeof(v) + 2;
+}
+inline void denc_varint_lowz(uint64_t v, bufferlist::contiguous_appender& p) {
+  int lowznib = v ? (ctz(v) / 4) : 0;
+  if (lowznib > 3)
+    lowznib = 3;
+  v >>= lowznib * 4;
+  v <<= 2;
+  v |= lowznib;
+  denc_varint(v, p);
+}
+
+template<typename T>
+inline void denc_varint_lowz(T& v, bufferptr::iterator& p)
+{
+  uint64_t i;
+  denc_varint(i, p);
+  int lowznib = (i & 3);
+  i >>= 2;
+  i <<= lowznib * 4;
+  v = i;
+}
+
+// signed varint + lowz encoding
+//
+// first low bit = 1 for negative, 0 for positive
+// next 2 bits = how many low zero bits (nibbles)
+// high bit of each byte = another byte follows
+// (so, 4 bits data in first byte, 7 bits data thereafter)
+inline void denc_signed_varint_lowz(int64_t v, size_t& p) {
+  p += sizeof(v) + 2;
+}
+inline void denc_signed_varint_lowz(int64_t v,
+				    bufferlist::contiguous_appender& p) {
+  bool negative = false;
+  if (v < 0) {
+    v = -v;
+    negative = true;
+  }
+  int lowznib = v ? (ctz(v) / 4) : 0;
+  if (lowznib > 3)
+    lowznib = 3;
+  v >>= lowznib * 4;
+  v <<= 3;
+  v |= lowznib << 1;
+  v |= (int)negative;
+  denc_varint(v, p);
+}
+
+template<typename T>
+inline void denc_signed_varint_lowz(T& v, bufferptr::iterator& p)
+{
+  int64_t i;
+  denc_varint(i, p);
+  int lowznib = (i & 6) >> 1;
+  if (i & 1) {
+    i >>= 3;
+    i <<= lowznib * 4;
+    v = -i;
+  } else {
+    i >>= 3;
+    i <<= lowznib * 4;
+    v = i;
+  }
+}
+
+
+// LBA
+//
+// first 1-3 bits = how many low zero bits
+//     *0 = 12 (common 4 K alignment case)
+//    *01 = 16
+//   *011 = 20
+//   *111 = byte
+// then 28-30 bits of data
+// then last bit = another byte follows
+// high bit of each subsequent byte = another byte follows
+inline void denc_lba(uint64_t v, size_t& p) {
```

Comment:
Sounds like a job for a featured type.

Allen Samuels
SanDisk |a Western Digital brand
2880 Junction Avenue, Milpitas, CA 95134
T: +1 408 801 7030| M: +1 408 780 6416
allen.samuels@SanDisk.com<mailto:allen.samuels@SanDisk.com>

From: Sage Weil [mailto:notifications@github.com]
Sent: Wednesday, September 14, 2016 8:41 AM
To: ceph/ceph <ceph@noreply.github.com>
Cc: Allen Samuels <Allen.Samuels@sandisk.com>; Comment <comment@noreply.github.com>
Subject: Re: [ceph/ceph] RFC: fast denc encoding (#11027)


In src/include/denc.h<https://github.com/ceph/ceph/pull/11027#discussion_r78775041>:

> +    v = i;

> +  }

> +}

> +

> +

> +// LBA

> +//

> +// first 1-3 bits = how many low zero bits

> +//     *0 = 12 (common 4 K alignment case)

> +//    *01 = 16

> +//   *011 = 20

> +//   *111 = byte

> +// then 28-30 bits of data

> +// then last bit = another byte follows

> +// high bit of each subsequent byte = another byte follows

> +inline void denc_lba(uint64_t v, size_t& p) {

The encoding for the type would then be fixed (always space efficient), whereas currently the containing type gets to decide. That is probably okay for bluestore though... i'll add it to trello.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/ceph/ceph/pull/11027/files/55ef06be7ba874cc1d8f085534a20baf4bc3a082#r78775041>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AFcGFFkJtjQYZevWXzW7p_0sA-kOlJy-ks5qqBVzgaJpZM4J4jDG>.